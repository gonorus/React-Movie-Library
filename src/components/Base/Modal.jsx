import React, { useEffect } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import './Modal.scss';

export const DisplayModal = (Component) => {
  const rooElement = document.getElementById('root');
  const modalContainer = document.createElement('div');
  modalContainer.className = 'modal-container';

  if (rooElement) {
    rooElement.appendChild(modalContainer);
    ReactDOM.render(Component, modalContainer);
  }
};

const Modal = ({ children }) => {
  useEffect(() => () => {
    const containerElement = document.getElementsByClassName('modal-container')[0];
    containerElement.remove();
  }, []);

  const OnCloseModal = () => {
    const containerElement = document.getElementsByClassName('modal-container')[0];
    if (containerElement) ReactDOM.unmountComponentAtNode(containerElement);
  };

  return (
    <div className='modal-dialog'>
      <button type='button' onClick={OnCloseModal}>Close</button>
      {children}
    </div>
  );
};
Modal.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired
};
export default Modal;
