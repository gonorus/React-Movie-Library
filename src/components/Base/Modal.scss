.modal-container {
  width: 100%;
  height: 100%;
  position: absolute;
  background-color: rgba(0,0,0,0.6);
  top: 0;
  left: 0;
  display: grid;
  place-content: center;

  .modal-dialog {
    background-color: white;
    padding: 8px;
    border-radius: .2em;
    display: grid;
    row-gap: 8px;

    button {
      align-self: center;
    }
  }
}