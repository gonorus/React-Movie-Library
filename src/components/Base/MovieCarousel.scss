.movie-carousel {
  width: 100%;
  display: flex;
  gap: 8px;
  padding: 4px;
  overflow-y: hidden;
  box-sizing: border-box;
}