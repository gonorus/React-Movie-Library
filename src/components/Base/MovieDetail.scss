.movie-details-container {
  max-width: 300px;
  max-height: 500px;
  padding: 16px;
  display: grid;
  row-gap: 8px;
  overflow: auto;

  img {
    max-height: 300px;
    max-width: 200px;
  }

  p {
    margin: 0;
  }

  .showbox {
    position: relative;
  }

  .movie-info {
    pre {
      white-space: pre-wrap;
    }
    
    p {
      text-align: justify;
    }
  }

  &::-webkit-scrollbar {
    height: 6px;
    width: 6px;
    background: #FFF;
  }

  &::-webkit-scrollbar-thumb {
    background: #393812;
    -webkit-border-radius: 1em;
  }
}