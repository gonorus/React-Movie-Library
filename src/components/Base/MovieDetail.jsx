import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import './MovieDetail.scss';

import Modal from './Modal';
import Spinner from '../Misc/Spinner';
import { GetRequest } from '../../helpers/HttpRequest';
import { API_URL } from '../../configurations/ApiConfig';
import Picture from './Picture';
import { DateToMDY } from '../../helpers/Date';

/**
 * displaying Loader message
 */
const MovieLoader = () => (
  <>
    <Spinner />
    <p>Please wait...</p>
  </>
);

/**
 * displaying the movie information
 * @param {*} props
 */
export const MovieInfo = ({ movie }) => {
  const {
    Title, ReleaseDate, ImagePath, Overview
  } = movie;
  return (
    <>
      <Picture src={ImagePath} />
      <div className='movie-info'>
        <pre>
          Title:
          {' '}
          {Title}
          {'\n'}
          Release:
          {' '}
          {DateToMDY(ReleaseDate)}
        </pre>
        <p>
          {Overview}
        </p>
      </div>
    </>
  );
};
MovieInfo.propTypes = {
  movie: PropTypes.shape({
    Title: PropTypes.string.isRequired,
    ReleaseDate: PropTypes.objectOf(Date).isRequired,
    ImagePath: PropTypes.string.isRequired,
    Overview: PropTypes.string.isRequired,
  }).isRequired,
};

/**
 * Movie Details dialog
 * @param {*} props
 * @returns
 */
const MovieDetail = ({ url }) => {
  const [movieDetails, setMovieDetails] = useState({});
  const [isLoaded, setIsLoaded] = useState(false);

  useEffect(() => {
    GetRequest(url)
      .then(
        (response) => {
          const {
            id, title, release_date, poster_path, // Movie Props
            name, first_air_date, // TV Props
            overview,
          } = response.data;
          setMovieDetails(({
            Id: id.toString(),
            Title: title || name,
            // eslint-disable-next-line camelcase
            ReleaseDate: release_date ? new Date(release_date) : new Date(first_air_date),
            ImagePath: API_URL.THUMBNAIL_IMAGE_URL.replace('/{ImagePath}', poster_path),
            Overview: overview,
          }));
        }
      )
      .catch(() => setMovieDetails(null))
      .finally(() => setIsLoaded(true));
  }, []);

  return (
    <Modal>
      <div className='movie-details-container'>
        {isLoaded === false && <MovieLoader />}
        {isLoaded === true && <MovieInfo movie={movieDetails} />}
      </div>
    </Modal>
  );
};
MovieDetail.propTypes = {
  url: PropTypes.string.isRequired
};
export default MovieDetail;
