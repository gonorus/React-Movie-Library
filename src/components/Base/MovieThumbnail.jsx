import React from 'react';
import PropTypes from 'prop-types';
import './MovieThumbnail.scss';

import Picture from './Picture';
import { DateToMDY } from '../../helpers/Date';
import { DisplayModal } from './Modal';
import MovieDetail from './MovieDetail';

/**
 * Movie Thumbnail with title
 */
const MovieThumbnail = ({ movie }) => {
  const {
    Title, ReleaseDate, ImagePath, MovieDetails
  } = movie;

  /**
   * should displaying Movie details
   */
  const OnMovieSelected = () => {
    DisplayModal(<MovieDetail url={MovieDetails} />);
  };

  /**
   * handle Enter Key pressed
   * @param {Event} evt
   */
  const OnKeyPressed = (evt) => {
    if (evt.charCode === 13) {
      OnMovieSelected();
    }
  };

  return (
    <div
      className='movie-thumbnail-card'
      role='button'
      tabIndex={0}
      onClick={OnMovieSelected}
      onKeyPress={OnKeyPressed}
    >
      <Picture src={ImagePath} alt={Title} />
      <div className='movie-thumbnail-info'>
        <h3>{Title}</h3>
        <p>{DateToMDY(ReleaseDate)}</p>
      </div>
    </div>
  );
};
MovieThumbnail.propTypes = {
  movie: PropTypes.shape({
    Title: PropTypes.string.isRequired,
    ReleaseDate: PropTypes.objectOf(Date).isRequired,
    ImagePath: PropTypes.string.isRequired,
    MovieDetails: PropTypes.string.isRequired,
  }).isRequired,
};
export default MovieThumbnail;
