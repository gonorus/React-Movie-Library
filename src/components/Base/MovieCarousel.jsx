import React from 'react';
import PropTypes from 'prop-types';
import './MovieCarousel.scss';

import MovieThumbnail from './MovieThumbnail';

/**
 * @param {*} props
 * @returns
 */
const MovieCarousel = ({ movies }) => (
  <div className='movie-carousel'>
    {
      movies.map((movie) => <MovieThumbnail key={movie.Id} movie={movie} />)
    }
  </div>
);
MovieCarousel.propTypes = {
  movies: PropTypes.arrayOf(PropTypes.shape({
    Id: PropTypes.string.isRequired,
    Title: PropTypes.string.isRequired,
    ReleaseDate: PropTypes.objectOf(Date).isRequired,
    ImagePath: PropTypes.string.isRequired,
  })),
};
MovieCarousel.defaultProps = {
  movies: [
    {
      Id: '1', Title: 'Movie Title', ReleaseDate: new Date(), ImagePath: 'http://localhost/example.jpg',
    },
    {
      Id: '2', Title: 'Movie Title', ReleaseDate: new Date(), ImagePath: 'https://www.themoviedb.org/t/p/w440_and_h660_face/jTswp6KyDYKtvC52GbHagrZbGvD.jpg',
    },
    {
      Id: '3', Title: 'Movie Title', ReleaseDate: new Date(), ImagePath: 'http://localhost/example.jpg',
    },
    {
      Id: '4', Title: 'Movie Title', ReleaseDate: new Date(), ImagePath: 'https://www.themoviedb.org/t/p/w440_and_h660_face/jTswp6KyDYKtvC52GbHagrZbGvD.jpg',
    },
    {
      Id: '5', Title: 'Movie Title', ReleaseDate: new Date(), ImagePath: 'http://localhost/example.jpg',
    },
    {
      Id: '6', Title: 'Movie Title', ReleaseDate: new Date(), ImagePath: 'https://www.themoviedb.org/t/p/w440_and_h660_face/jTswp6KyDYKtvC52GbHagrZbGvD.jpg',
    },
    {
      Id: '7', Title: 'Movie Title', ReleaseDate: new Date(), ImagePath: 'http://localhost/example.jpg',
    },
    {
      Id: '8', Title: 'Movie Title', ReleaseDate: new Date(), ImagePath: 'https://www.themoviedb.org/t/p/w440_and_h660_face/jTswp6KyDYKtvC52GbHagrZbGvD.jpg',
    },
    {
      Id: '9', Title: 'Movie Title', ReleaseDate: new Date(), ImagePath: 'http://localhost/example.jpg',
    },
    {
      Id: '10', Title: 'Movie Title', ReleaseDate: new Date(), ImagePath: 'https://www.themoviedb.org/t/p/w440_and_h660_face/jTswp6KyDYKtvC52GbHagrZbGvD.jpg',
    },
  ],
};
export default MovieCarousel;
