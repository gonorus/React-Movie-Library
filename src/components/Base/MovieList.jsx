import React, { useEffect, useState } from 'react';

import { API_URL, MOVIE_DB, TV_DB } from '../../configurations/ApiConfig';
import { GetRequest } from '../../helpers/HttpRequest';
import MovieCarousel from './MovieCarousel';

/**
 * extracting the movie list response into derived Movie object
 * @param {*} dataResult
 * @returns
 */
const ExtractMovieList = (dataResult) => dataResult.map(
  ({
    id, title, release_date, poster_path, // Movie Props
    name, first_air_date, // TV Props
  }) => ({
    Id: id.toString(),
    Title: title || name,
    // eslint-disable-next-line camelcase
    ReleaseDate: release_date ? new Date(release_date) : new Date(first_air_date),
    ImagePath: API_URL.THUMBNAIL_IMAGE_URL.replace('/{ImagePath}', poster_path),
    MovieDetails: title ? MOVIE_DB.DETAILS.replace('{Id}', id) : TV_DB.DETAILS.replace('{Id}', id),
  })
);

export const PopularTvShows = () => {
  const [Movies, setMovies] = useState([]);

  useEffect(() => {
    GetRequest(TV_DB.POPULAR)
      .then(
        ({ data }) => {
          const result = ExtractMovieList(data.results);
          if (result.length > 0) setMovies(result);
          else setMovies(null);
        }
      ).catch(() => setMovies(null));
  }, []);

  return Movies ? <MovieCarousel title='Popular TV Show' movies={Movies} /> : <></>;
};

export const PopularMovies = () => {
  const [Movies, setMovies] = useState([]);

  useEffect(() => {
    GetRequest(MOVIE_DB.POPULAR)
      .then(
        ({ data }) => {
          const result = ExtractMovieList(data.results);
          if (result.length > 0) setMovies(result);
          else setMovies(null);
        }
      ).catch(() => setMovies(null));
  }, []);

  return Movies ? <MovieCarousel title='Popular Movies' movies={Movies} /> : <></>;
};

export const NowPlayingMovies = () => {
  const [Movies, setMovies] = useState([]);

  useEffect(() => {
    GetRequest(MOVIE_DB.NOW_PLAYING)
      .then(
        ({ data }) => {
          const result = ExtractMovieList(data.results);
          if (result.length > 0) setMovies(result);
          else setMovies(null);
        }
      ).catch(() => setMovies(null));
  }, []);

  return Movies ? <MovieCarousel title='Now Playing Movies' movies={Movies} /> : <></>;
};

export const UpcomingMovies = () => {
  const [Movies, setMovies] = useState([]);

  useEffect(() => {
    GetRequest(MOVIE_DB.UPCOMING)
      .then(
        ({ data }) => {
          const result = ExtractMovieList(data.results);
          if (result.length > 0) setMovies(result);
          else setMovies(null);
        }
      ).catch(() => setMovies(null));
  }, []);

  return Movies ? <MovieCarousel title='Upcoming Movies' movies={Movies} /> : <></>;
};
