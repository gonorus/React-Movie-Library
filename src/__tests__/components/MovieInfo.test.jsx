import React from 'react';
import renderer from 'react-test-renderer';
import { MovieInfo } from '../../components/Base/MovieDetail';

test('Movie info displayed', () => {
  const movie = {
    Title: 'Movie Title',
    ReleaseDate: new Date('2021-12-04'),
    ImagePath: 'http://localhost/example.jpg',
    Overview: 'movie overview',
  };
  const component = renderer.create(
    <MovieInfo movie={movie} />
  );
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
