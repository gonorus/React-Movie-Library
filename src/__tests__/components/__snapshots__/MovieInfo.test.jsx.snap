// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Movie info displayed 1`] = `
Array [
  <div
    className="picture-container"
  >
    <img
      loading="lazy"
      onError={[Function]}
      onLoad={[Function]}
      src="http://localhost/example.jpg"
    />
    <div
      className="showbox"
    >
      <div
        className="loader"
        style={
          Object {
            "width": 50,
          }
        }
      >
        <svg
          className="circular"
          viewBox="25 25 50 50"
        >
          <circle
            className="path"
            cx="50"
            cy="50"
            fill="none"
            r="20"
            strokeMiterlimit="10"
            strokeWidth="2"
          />
        </svg>
      </div>
    </div>
  </div>,
  <div
    className="movie-info"
  >
    <pre>
      Title:
       
      Movie Title
      

      Release:
       
      Dec 4, 2021
    </pre>
    <p>
      movie overview
    </p>
  </div>,
]
`;
